whitesource/util.py

```diff
@@ -1,10 +1,150 @@
+    above = _sort_struct(above)
+
+    ttable_header = ('Component', 'Greatest CVSS-V3', 'Corresponding CVE')
+    ttables = []
+
+    for source in above, below:
+        if len(source) == 0:
+            ttables.append("")
+            continue
+        ttable_data = (
+            (
+                component,
+                dic["CVSS-V3"],
+                dic["CVE"]
+            ) for component, dic, in source
+        )
+
+        ttable = tabulate.tabulate(
+            headers=ttable_header,
+            tabular_data=ttable_data,
+            tablefmt=tablefmt,
+            colalign=('left', 'center', 'center')
+        )
+
+        ttables.append(ttable)
+
+    return ttables
+
+
+def assemble_mail_body(tables: list,
+                       threshold: float):
+    return f'''
+        <div>
+            <p>
+                Note: you receive this E-Mail, because you were configured as a mail recipient
+                (see .ci/pipeline_definitions)
+                To remove yourself, search for your e-mail address in said file and remove it.
+            </p>
+            <br></br>
+            <p>
+                The following component(s) have a CVSS-V3 greater than the configured threshold of
+                {threshold}. It is configured at the
+                <a href="https://github.wdf.sap.corp/kubernetes/cc-config">
+                    pipeline definition
+                </a>.
+            </p>
+            {tables[0]}
+            <br></br>
+            <br></br>
+            <p>
+                These are the remaining component(s) with a CVSS-V3 lower than {threshold}
+            </p>
+            {tables[1]}
+            <br></br>
+            <br></br>
+            <p>
+                WhiteSource triage has to be done on the
+                <a href="https://saas.whitesourcesoftware.com/Wss/WSS.html#!alertsReport">
+                    WhiteSource Alert Reporting
+                </a>
+                page. Appropriate filters have to applied manually, "Gardener" is a matching keyword.
+            </p>
+        </div>
+    '''
+
+
+def find_greatest_cve(projects,
+                      client):
+    report = {}
+
+    # get all projects for product
+    for project in projects["projects"]:
+        pname = project["projectName"]
+        ptoken = project["projectToken"]
+
+        # get vulnerability report per project
+        ci.util.info(f'retrieving project vulnerability report for {pname}')
+        pvr = client.get_project_vulnerability_report(project_token=ptoken)
+
+        # find greatest cve per project
+        for vul in json.loads(pvr.content)["vulnerabilities"]:
+            try:
+                if report.get(pname) is None or vul["cvss3_score"] > report[pname]["CVSS-V3"]:
+                    report[pname] = {
+                        "CVSS-V3": vul["cvss3_score"],
+                        "CVE": vul["name"],
+                    }
+            except KeyError:
+                # WS are ignored
```

Comment:
what's WS?
in which case will those key-errs occur, and why is it okay to ignore them silently?

I know, it is a pythonic pattern to "ask for forgiveness"; however, in this case, there is quite some risk that this will hide actual implementation errors, as there are a lot of dict-element-accesses (btw: did I already mention dataclasses and dacite do a great job in deserialising dicts?

Reply:
'Whereas the NVD publishes security vulnerabilities using a "CVE-" prefixed identifier, WhiteSource classifies non-NVD security vulnerabilities using a "WS-" prefix' - https://whitesource.atlassian.net/wiki/spaces/WD/pages/470188110/Notice+Regarding+CVE+and+WS+Vulnerability+Identifiers

Unfortunately the `getProjectVulnerabilityReport` from the WhiteSource API includes some inconsistency. While all `CVE*` entries have a key called `cvss3_score` some `WS*` entries call the corresponding key either `cvss3_score` or just `score`. I adjusted the logic to look for both possibilities.